### ISSUE TITLE
IDL sequence should be Iterable, not Array

### ISSUE THREAD

**Issue by kainino0x** [darionco/bikeshed-to-ts#6]
WebIDL `sequence`s are (I think) equivalent to TypeScript `Iterable`s - the WebIDL conversion rules from JS will iterate the value to produce an array. For example:
```webidl
dictionary GPUBindGroupLayoutDescriptor : GPUObjectDescriptorBase {
    required sequence<GPUBindGroupLayoutEntry> entries;
};
```
generates:
```ts
  interface GPUBindGroupLayoutDescriptor extends GPUObjectDescriptorBase {
    entries: Array<GPUBindGroupLayoutEntry>;
  }
```
but should generate:
```ts
  interface GPUBindGroupLayoutDescriptor extends GPUObjectDescriptorBase {
    entries: Iterable<GPUBindGroupLayoutEntry>;
  }
```

**Comment by darionco** [comment]
The [WebIDL spec](https://heycam.github.io/webidl/#es-sequence) defines a `sequence` as an ECMAScript `Array`.

Unfortunately `Array` and `Iterable` are not a 1 to 1 mapping because `Array` always allows index access to its members through square brackets notation while an `Iterable` is only expected to implement the `Symbol.iterator` property as seen in the [TypeScript declaration](https://github.com/microsoft/TypeScript/blob/131875bb849c0a9c56c26d09881453aace0cbfed/lib/lib.es2015.iterable.d.ts#L50)

An alternative would be declaring an iterable object as defined in the [WebIDL spec](https://heycam.github.io/webidl/#idl-iterable).

**Comment by kainino0x** [comment]
Ah, there's an effective difference between _input_ `sequence`s and _output_ `sequence`s.
IDL allows any JS iterable to be passed _in_ to a `sequence`-typed WebIDL binding, but when it passes a WebIDL `sequence` _out_ to a JS value it produces an array.

WebGPU doesn't actually use any output `sequence`s (IIRC), instead using `FrozenArray` (which as of today it only uses in one place).

**Comment by kainino0x** [comment]
Unfortunately I can't think of an easy way to be 100% correct about input vs output types. Many types can be used in both input and output (any that are sequence or dictionary, at least). IIRC these are only _recommended_ to be used in input situations but they're _allowed_ to be used in output situations as well.

**Comment by darionco** [comment]
yeah, this one is a difficult one.

for function arguments or return types a simple input/output heuristic could be used:
```webidl
undefined setSequence(sequence<any> value);
sequence<any> getSequence();
```
would translate to:
```ts
setSequence(Iterable<any> value): void;
getSequence(): Array<any>;
```

Properties are more difficult since they can either be `readonly` or `read/write` which makes it so they must always be `Array`.

The closest solution that I can think of is to follow the heuristic above and create setters and getters to be verbose:
```webidl
dictionary GPUBindGroupLayoutDescriptor : GPUObjectDescriptorBase {
    required sequence<GPUBindGroupLayoutEntry> entries;
};
```
would translate to:
```ts
interface GPUBindGroupLayoutDescriptor extends GPUObjectDescriptorBase {
    set entries(value: Iterable<GPUBindGroupLayoutEntry>);
    get entries(): Array<GPUBindGroupLayoutEntry>;
}
  ```
  
  but that feels off

**Comment by kainino0x** [comment]
Yeah, that wouldn't work if the type were used as a local variable:

```ts
const x: GPUBindGroupLayoutDescriptor = { entries: new Set(...) };
const y = x.entries[0];
```

Unfortunately the only thing I can think of is to generate both input and output versions of every type (perhaps differentiated by a type parameter??), then optionally prune any input or output variants that go unused...

**Comment by kainino0x** [comment]
We can not worry about this for now for WebGPU though. The diff that we have to deal with manually is very small.

**Comment by kainino0x** [comment]
In principle all the descriptors need to be `readonly` on input too, but fortunately (I guess) typescript is not strict about this. https://github.com/microsoft/TypeScript/issues/18770

I did a little experimentation with the idea I had above and it _could_ work:
https://github.com/gpuweb/types/compare/main...kainino0x:idlsequence-proof-of-concept

**Comment by kainino0x** [comment]
https://github.com/toji/webidl2ts/pull/1 will sort of fix this, at least it will get rid of the manual changes we have to make to webgpu's generated types. We can keep considering on https://github.com/gpuweb/types how to improve the things discussed above.

**Comment by kainino0x** [comment]
That simple fix works because:

> WebGPU doesn't actually use any output `sequence`s (IIRC), instead using `FrozenArray` (which as of today it only uses in one place).
